File: backend/app/schemas/platform_limits.py
from pydantic import BaseModel

from app.models.listing import ListingPlatform
from app.services.platform_limits import LimitField, Severity


class FieldViolation(BaseModel):
    """One value that breaches one limit.

    `message` carries the complete human sentence rather than leaving the frontend to
    assemble one from the parts — the house style (services/variants.py) is that the
    backend names the offending thing, the limit and the platform in one go, so the same
    wording appears everywhere it surfaces. The structured fields are there for sorting,
    filtering and highlighting, not for rebuilding the sentence.
    """

    field: LimitField
    severity: Severity
    current_value: str
    current_length: int | None
    limit: str
    imposed_by: ListingPlatform
    message: str
    # Present only where a mechanical fix is unambiguous. None means a human decides —
    # notably for every blocker, where "just truncate it" would be wrong.
    suggested_value: str | None


class UnitCompatibility(BaseModel):
    """Violations for one variant (or for the product itself when it has none), matching
    how listing_sync partitions a product into checkable units."""

    variant_id: int | None
    variant_name: str | None
    sku: str | None
    violations: list[FieldViolation]


class ProductCompatibility(BaseModel):
    product_id: int
    product_name: str
    product_sku: str | None
    # True when anything on the product or its units would stop a listing being created.
    is_blocked: bool
    violations: list[FieldViolation]
    units: list[UnitCompatibility]


class CatalogueCompatibilityReport(BaseModel):
    """Summary counts with per-product drill-down.

    `products` holds only the entries with something to report; `total_products` is the
    whole active catalogue, so the UI can say "3 of 26" without a second call.
    """

    platform: ListingPlatform
    total_products: int
    blocked_count: int
    warning_count: int
    products: list[ProductCompatibility]

File: frontend/src/api/platformLimits.ts
import { api } from "./client";
import type { ListingPlatform } from "./types";

export type LimitField =
  | "sku_max_length"
  | "title_max_length"
  | "title_charset"
  | "description_max_length"
  | "variation_attribute_max_count"
  | "variation_max_count"
  | "attribute_name_max_length"
  | "attribute_value_max_length"
  | "attribute_value_charset"
  | "image_max_count"
  | "price_decimal_places"
  | "quantity_max";

export type ViolationSeverity = "blocker" | "warning";

// `message` is the complete sentence from the backend, which already names the value, the
// limit and the platform that imposes it. Render it as-is rather than reassembling one
// from the parts — the same wording then appears wherever a violation surfaces.
export interface FieldViolation {
  field: LimitField;
  severity: ViolationSeverity;
  current_value: string;
  current_length: number | null;
  limit: string;
  imposed_by: ListingPlatform;
  message: string;
  suggested_value: string | null;
}

export interface UnitCompatibility {
  variant_id: number | null;
  variant_name: string | null;
  sku: string | null;
  violations: FieldViolation[];
}

export interface ProductCompatibility {
  product_id: number;
  product_name: string;
  product_sku: string | null;
  is_blocked: boolean;
  violations: FieldViolation[];
  units: UnitCompatibility[];
}

// `products` holds only the entries with something to report; `total_products` is the
// whole active catalogue, so "3 of 26" needs no second call.
export interface CatalogueCompatibilityReport {
  platform: ListingPlatform;
  total_products: number;
  blocked_count: number;
  warning_count: number;
  products: ProductCompatibility[];
}

export const platformLimitsApi = {
  catalogueCompatibility: (platform: ListingPlatform) =>
    api.get<CatalogueCompatibilityReport>(`/platforms/${platform}/catalogue-compatibility`),
};
